cogs/birthday/birthday.py

```diff
@@ -519,46 +550,111 @@ async def setSelfBirthday(self, ctx: Context, *, birthday: MonthDayConverter):
+                    ctx.author.discriminator,
+                    ctx.author.id,
+                    birthdayStr,
                 )
+
+                # explicitly check to see if user should be added to role, if the month
+                # and day just so happen to be the same as it is now.
+                await self.checkBirthday()
                 return
+            except discord.Forbidden:
+
+                def check(msg: discord.Message):
+                    return msg.author == ctx.author and msg.channel == ctx.channel
+
+                await ctx.send(
+                    "You have disabled DMs from this server. Would you "
+                    "still like to continue here?"
+                    f"Type {bold('`yes`', escape_formatting=False)} to confirm."
+                )
+                try:
+                    response = await self.bot.wait_for("message", timeout=30.0, check=check)
+                except asyncio.TimeoutError:
+                    await ctx.send(f"{headerBad}: You took too long. Not setting your birthday.")
+                    return
+
+                if response.content.lower() != "yes":
+                    await ctx.send(f"{headerBad}: Declined. Not setting your birthday.")
+                    return
+
+            msg: discord.Message = await ctx.send(f"{headerWarn}: {confirmationStr}")
+            try:
+                response = await self.bot.wait_for("message", timeout=10.0, check=check)
+            except asyncio.TimeoutError:
+                try:
+                    await ctx.send(f"{headerBad}: No response. Aborting.")
+                    await msg.delete()
+                    return
+                except (discord.NotFound, discord.HTTPException, discord.Forbidden) as e:
+                    self.logger.debug(e)
+                    return
 
             if response.content.lower() != "yes":
-                await ctx.author.send(f"{headerBad}: Declined. Not setting your birthday.")
-                return
+                await ctx.send(
+                    f"{headerBad}: Declined. Not setting your birthday.",
+                )
+                try:
+                    await msg.delete()
+                    return
+                except (discord.NotFound, discord.HTTPException, discord.Forbidden) as e:
+                    self.logger.debug(e)
+                    return
 
             await birthdayConfig.get_attr(KEY_BDAY_MONTH).set(birthday.month)
             await birthdayConfig.get_attr(KEY_BDAY_DAY).set(birthday.day)
             await birthdayConfig.get_attr(KEY_ADDED_BEFORE).set(True)
-
-            await ctx.author.send(
+            await ctx.send(
                 f"{headerGood}: Successfully set your birthday to "
-                f"{spoiler(bold(birthdayStr, escape_formatting=False), escape_formatting=False)}."
+                f"{spoiler(bold(birthdayStr, escape_formatting=False), escape_formatting=False)}.",
+                delete_after=5,
             )
```

Comment:
The logic here is pretty much repeating the work in the outer `try` block, except here (not in DMs) we try to delete messages containing birthdates (whereas in the outer `try` block we needn't care much 'cause it's sent in DMs).

Can we refactor so that we don't repeat a big portion of code like this? After agreeing to proceed in a server text channel, the general logic (specifically flow of prompting questions and expecting answers) doesn't differ from that in the DM case, so we can have a preprocessing step like this:
```py
def mainFlow(destination: Union[Context, User], carefree: bool = False):
    # param carefree indicates if we aren't gonna care about deleting msgs containing privacy-sensitive info (e.g., birthday information)

    # prompt
    try:
        await destination.send(f"{headerWarn}: {confirmationStr}")
    except discord.Forbidden:
        # here we fail at the first msg sent to `destination`
        # meaning we can't continue the flow anymore
        # so raise this exception again for the caller to handle
        raise
    # expect answer
    try:
        response = self.bot.wait_for("message")
    except: asyncio.TimeoutError:
        destination.send("No response. Aborting.")
    # answer check
    if response.content.lower() != "yes":
        destination.send("Not proceeding.")
    # proceed
    setBday() # basically means setting birth month, year etc. as needed
    sensitiveMsgTtl = None if carefree else 10.0 # sensitive message's time to live - the time for `delete_after`
    await destination.send(privateInfo, delete_after=sensitiveMsgTtl)
    # or
    msg = await destination.send(privateInfo)
    if not carefree:
        # wait for some time
        try:
            msg.delete()
        except Exception as e:
            self.logger.debug("%s", e)
            # handle exception appropriately
```
And then, on the outer scope, we can just invoke the `mainFlow`:
```py
try:
    mainFlow(ctx.author, carefree=True)
except discord.Forbidden: # DMs blocked
    # ask to proceed in server text channel
    # then, if OK
    mainFlow(ctx, carefree=False)
```

If it's worth, this can also be good for the logic in `getSelfBirthday` though I don't feel `getSelfBirthday` has a considerable or significant portion of repeated logic.

Reply:
+1

Reply:
I have implemented a main flow function as you described above, let me know if something looks off